CIP-0071/README.md

```diff
@@ -0,0 +1,246 @@
+  tx: Tx = ctx.tx;
+  minted_policy: MintingPolicyHash = ctx.get_current_minting_policy_hash();
+  redeemer.switch {
+    Mint => {
+      polls_are_still_open(tx.time_range)
+        && assets_were_spent(tx.minted, minted_policy, tx.outputs)
+        && assets_locked_in_script(tx, tx.minted)
+    },
+    // Burn code elided for space...
+  }
+}
+```
+
+#### "Vote" -> eUTxO Datum
+
+To cast the vote, the user sends the ballot NFT just created to a "ballot box".  Note that for reasons specified in [the "attacks" section below](#Creation-of-Ballot-Without-Casting-a-Vote) this needs to occur during the same transaction that the ballot was minted in.
+
+The datum is a simple object representing the voter who cast the vote and the vote itself:
+
+```ts
+type VoteDatum = {
+  voter: PubKeyHash,
+  vote: object
+};
+```
+
+The `voter` element is extremely important in this datum so that we know who minted the ballot NFT and who we should return it to.  At the end of the ballot counting process, this user will receive their ballot NFT back.
+
+Note that we are trying to avoid being overly prescriptive here with the specific vote type as we want the only limitations on the vote type to be those imposed by Cardano.  Further iterations of this standard should discuss the potential for how to implement ranked-choice voting (RCV) inside of this `vote` object, support multiple-choice vote selection, and more.
+
+#### "Ballot Box" -> Smart Contract
+
+Essentially, the "ballot box" is a smart contract with arbitrary logic decided upon by the authorized vote counter.  Some examples include:
+
+1. A ballot box that can be redeemed at any time by a tx signed by the authorized vote counter
+2. A ballot box that can be redeemed only after polls close
+2. A ballot box that can be redeemed once a majority of voters have sent in a ballot
+3. A ballot box that can be redeemed only by the specific wallet specified in the `voter` datum of each UTxO
+
+Because the ballot creation and vote casting process has already occurred on-chain we want to provide the maximum flexibility in the protocol here so that each project can decide what is best for their own community.  Helios code for the simple case enumerated as #1 above would be:
+
+```ts
+const EXPECTED_SIGNER: PubKeyHash = PubKeyHash::new(#0123456789abcdef)
+
+func main(ctx: ScriptContext) -> Bool {
+  ctx.tx.is_signed_by(EXPECTED_SIGNER)
+}
+```
+
+### The Vote Counting Process
+
+#### "Ballot Counter" -> Authorized Wallet
+
+Given the flexible nature of the ["ballot box" smart contract](#Ballot-Box---Smart-Contract) enumerated above, we propose a simple algorithm for counting votes and returning the ballots to the user:
+
+1. Ensure the polls are closed (can be either on or off-chain)
+2. Iterate through all UTxOs in forward-time-order locked in the "ballot box" and for each:
+	* Determine which assets are inside of that UTxO
+	* Mark their most recent vote to match the `vote` object in the UTxOs datum
```

Comment:
I think this last point answer a question I was about to ask about multiple votes. What happens when a voter mints multiple ballots and votes multiple time? This suggests only the most recent vote is counted.

Reply:
Correct.  Only the most recent vote is counted, although all votes are technically "valid". Basically we are allowing voters to change their mind.